File: Napakalaki_Ruby/specific_bad_consequence.rb
#encoding: utf-8

require_relative 'bad_consequence.rb'

module NapakalakiGame
  class SpecificBadConsequence < BadConsequence

    attr_reader :specificVisibleTreasures, :specificHiddenTreasures

    def initialize (aText, someLevels, someSpecificVisibleTreasures, someSpecificHiddenTreasures)
      super(aText, someLevels)
      @specificVisibleTreasures = Array.new(someSpecificVisibleTreasures)
      @specificHiddenTreasures = Array.new(someSpecificHiddenTreasures)
    end
    public_class_method :new

    def isEmpty
      (@specificVisibleTreasures.empty?) && (@specificHiddenTreasures.empty?)
    end

    def substractVisibleTreasure (t)
      for i in 0..@specificVisibleTreasures.length-1 do
        if t.type == @specificVisibleTreasures[i]
          @specificVisibleTreasures.delete_at(i)
          break
        end
      end
    end

    def substractHiddenTreasure (t)
      for i in 0..@specificHiddenTreasures.length-1 do
        if t.type == @specificHiddenTreasures[i]
          @specificHiddenTreasures.delete_at(i)
          break
        end
      end
    end

    def adjustToFitTreasureLists (v, h)
      auxVisible = Array.new(@specificVisibleTreasures)
      auxArray = Array.new(auxVisible)
      v.each do |treasure|
        for i in 0..auxArray.length-1 do
          if auxArray[i] == treasure.type
            auxArray.delete_at(i)
            break
          end
        end
      end
      auxArray.each do |tKind|
        auxVisible.delete_at(auxVisible.find_index(tKind))
      end

      auxHidden = Array.new(@specificHiddenTreasures)
      auxArray = Array.new(auxHidden)
      h.each do |treasure|
        for i in 0..auxArray.length-1 do
          if auxArray[i] == treasure.type
            auxArray.delete_at(i)
            break
          end
        end
      end
      auxArray.each do |tKind|
        auxHidden.delete_at(auxHidden.find_index(tKind))
      end

      newBadCons = SpecificBadConsequence.new("", 0, auxVisible, auxHidden)
    end

    def to_s
      super + "Specific Visible Treasures = " + @specificVisibleTreasures.to_s + "\nSpecific Hidden Treasures = " + @specificHiddenTreasures.to_s + "\n"
    end
  end
end

File: Napakalaki/src/NapakalakiGame/SpecificBadConsequence.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package NapakalakiGame;

import java.util.ArrayList;

/**
 *
 * @author francisco
 */
public class SpecificBadConsequence extends BadConsequence{
    private ArrayList<TreasureKind> specificVisibleTreasures;
    private ArrayList<TreasureKind> specificHiddenTreasures;
    
    public SpecificBadConsequence(String Text, int numLevels, ArrayList<TreasureKind> tVisible, 
                            ArrayList<TreasureKind> tHidden){
        super(Text, numLevels);
        specificVisibleTreasures = new ArrayList(tVisible);
        specificHiddenTreasures = new ArrayList(tHidden);
    }

    @Override
    public boolean isEmpty() {
        return specificVisibleTreasures.isEmpty() && specificHiddenTreasures.isEmpty();
    }

    @Override
    public void substractVisibleTreasure(Treasure t) {
        int arraysize = specificVisibleTreasures.size();
        for (int i=0; i<arraysize; i++){
            if (t.getType() == specificVisibleTreasures.get(i)){
                specificVisibleTreasures.remove(i);
                break;
            }
        }
    }
    
    @Override
    public void substractHiddenTreasure(Treasure t) {
        int arraysize = specificHiddenTreasures.size();
        for (int i=0; i<arraysize; i++){
            if (t.getType() == specificHiddenTreasures.get(i)){
                specificHiddenTreasures.remove(i);
                break;
            }
        }
    }

    @Override
    public SpecificBadConsequence adjustToFitTreasureLists(ArrayList<Treasure> v, ArrayList<Treasure> h) {
        int size;
        
        ArrayList<TreasureKind> auxVisible = new ArrayList(specificVisibleTreasures);
        ArrayList<TreasureKind> auxArray = new ArrayList(auxVisible);
        for (Treasure treasure : v){
            size = auxArray.size();
            for (int i=0; i<size; i++){
                if (auxArray.get(i) == treasure.getType()){
                    auxArray.remove(i);
                    break;
                }
            }
        }
        
        for (TreasureKind tKind : auxArray){
            auxVisible.remove(tKind);
        }
        
        ArrayList<TreasureKind> auxHidden = new ArrayList(specificHiddenTreasures);
        auxArray = new ArrayList(auxHidden);
        for (Treasure treasure : v){
            size = auxArray.size();
            for (int i=0; i<size; i++){
                if (auxArray.get(i) == treasure.getType()){
                    auxArray.remove(i);
                    break;
                }
            }
        }
        
        for (TreasureKind tKind : auxArray){
            auxHidden.remove(tKind);
        }

        SpecificBadConsequence newBadCons = new SpecificBadConsequence("", 0, auxVisible, auxHidden);
        
        return newBadCons;
    }
    
    @Override
    public String toString(){
        return super.toString() + "Specific Visible Treasures = " + specificVisibleTreasures.toString() +
                             "\nSpecific Hidden Treasures = " + specificHiddenTreasures.toString() + "\n";
    }
}
